include/aws/common/bigint.h

```diff
@@ -0,0 +1,144 @@
+#ifndef AWS_COMMON_BIGINT_H
+#define AWS_COMMON_BIGINT_H
+
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+
+#include <aws/common/array_list.h>
+#include <aws/common/byte_buf.h>
+
+/*
+ * A basic big integer implementation using 2^32 as the base.  Algorithms used are formalizations of the basic
+ * grade school operations everyone knows and loves (as formalized in AoCP Vol 2, 4.3.1).  Current use case
+ * targets do not yet involve a domain large enough that its worth exploring more complex algorithms.
+ */
+struct aws_bigint {
```

Comment:
What about `aws_bigint_is_valid()`? What are the set of invariants for this data structure? For example,
```
bool aws_bigint_is_valid(const struct aws_bigint * bigint) {
/* Pointer validity checks */
    if (!bigint) {
        return false;
    }
    bool digits_is_valid = aws_array_list_is_valid(&bigint->digits);
    bool sign_valid = (sign == 1 || sign == -1);
    return digits_is_valid && sign_valid;
}
```
It could be check as pre- and postconditions in every function of this module.

Reply:
Added an is_valid function and pre condition checks.  Post condition could be added on demand whenever proofs get added.

Reply:
Nice! You didn't added ` bool sign_valid = (sign == 1 || sign == -1);`, isn't necessary anymore?

Reply:
Actually I went and re-read the comment and updated is_valid to include it.